diff --git a/subprojects/gst-plugins-good/sys/v4l2/gstv4l2deviceprovider.c b/subprojects/gst-plugins-good/sys/v4l2/gstv4l2deviceprovider.c
index 5eeec7dc8b..5fb3edc86f 100644
--- a/subprojects/gst-plugins-good/sys/v4l2/gstv4l2deviceprovider.c
+++ b/subprojects/gst-plugins-good/sys/v4l2/gstv4l2deviceprovider.c
@@ -172,10 +172,6 @@ gst_v4l2_device_provider_probe_device (GstV4l2DeviceProvider * provider,
 
   if (caps == NULL)
     goto close;
-  if (gst_caps_is_empty (caps)) {
-    gst_caps_unref (caps);
-    goto close;
-  }
 
   device = gst_v4l2_device_new (device_path,
       device_name ? device_name : (gchar *) v4l2obj->vcap.card, caps, type,
