diff --git a/subprojects/gst-plugins-good/gst/matroska/matroska-demux.c b/subprojects/gst-plugins-good/gst/matroska/matroska-demux.c
index 058455be91..12b0d1f313 100644
--- a/subprojects/gst-plugins-good/gst/matroska/matroska-demux.c
+++ b/subprojects/gst-plugins-good/gst/matroska/matroska-demux.c
@@ -6515,6 +6515,37 @@ gst_matroska_demux_video_caps (GstMatroskaTrackVideoContext *
     caps = gst_video_info_to_caps (&info);
     *codec_name = gst_pb_utils_get_codec_description (caps);
     context->alignment = 32;
+  } else if (!strcmp (codec_id, GST_MATROSKA_CODEC_ID_VIDEO_RVL)) {
+    gint bpp=8;
+    GstVideoFormat format=GST_VIDEO_FORMAT_GRAY16_LE;
+
+    switch (videocontext->fourcc) {
+      case GST_MAKE_FOURCC ('1', ' ', ' ', ' '):
+        bpp = 1;
+        break;
+      case GST_MAKE_FOURCC ('2', ' ', ' ', ' '):
+        bpp = 2;
+        break;
+      case GST_MAKE_FOURCC ('4', ' ', ' ', ' '):
+        bpp = 4;
+        break;
+      case GST_MAKE_FOURCC ('8', ' ', ' ', ' '):
+        bpp = 8;
+        break;
+      case GST_MAKE_FOURCC ('1', '6', ' ', ' '):
+        bpp = 16;
+        break;
+      default:
+        GST_DEBUG ("Unknown fourcc %" GST_FOURCC_FORMAT,
+            GST_FOURCC_ARGS (videocontext->fourcc));
+        return NULL;
+    }
+    caps = gst_caps_new_simple ("video/rvl",
+        "format", G_TYPE_STRING, gst_video_format_to_string (format), NULL,
+        "bpp", G_TYPE_INT, bpp, NULL
+    );
+    context->intra_only = TRUE;
+    *codec_name = g_strdup ("RVL depth map");
   } else if (!strcmp (codec_id, GST_MATROSKA_CODEC_ID_VIDEO_MPEG4_SP)) {
     caps = gst_caps_new_simple ("video/x-divx",
         "divxversion", G_TYPE_INT, 4, NULL);
diff --git a/subprojects/gst-plugins-good/gst/matroska/matroska-ids.h b/subprojects/gst-plugins-good/gst/matroska/matroska-ids.h
index 3e65addd55..b5606221a3 100644
--- a/subprojects/gst-plugins-good/gst/matroska/matroska-ids.h
+++ b/subprojects/gst-plugins-good/gst/matroska/matroska-ids.h
@@ -381,6 +381,8 @@
 #define GST_MATROSKA_CODEC_ID_VIDEO_MPEGH_HEVC   "V_MPEGH/ISO/HEVC"
 #define GST_MATROSKA_CODEC_ID_VIDEO_PRORES       "V_PRORES"
 #define GST_MATROSKA_CODEC_ID_VIDEO_FFV1         "V_FFV1"
+/* Aivero-specific: RVL encoded depth map */
+#define GST_MATROSKA_CODEC_ID_VIDEO_RVL          "V_RVL"
 
 #define GST_MATROSKA_CODEC_ID_AUDIO_MPEG1_L1       "A_MPEG/L1"
 #define GST_MATROSKA_CODEC_ID_AUDIO_MPEG1_L2       "A_MPEG/L2"
diff --git a/subprojects/gst-plugins-good/gst/matroska/matroska-mux.c b/subprojects/gst-plugins-good/gst/matroska/matroska-mux.c
index 4780832de9..1855e97662 100644
--- a/subprojects/gst-plugins-good/gst/matroska/matroska-mux.c
+++ b/subprojects/gst-plugins-good/gst/matroska/matroska-mux.c
@@ -148,6 +148,10 @@ static GstStaticPadTemplate videosink_templ =
         "video/x-raw, "
         "format = (string) { YUY2, I420, YV12, UYVY, AYUV, GRAY8, BGR, RGB }, "
         COMMON_VIDEO_CAPS "; "
+        "video/rvl, "
+        "format = (string) { GRAY16_LE }, "
+        "bpp = (int) { 1, 2, 4, 8, 16 }, "
+        COMMON_VIDEO_CAPS "; "
         "video/x-prores, "
         COMMON_VIDEO_CAPS "; "
         "video/x-wmv, " "wmvversion = (int) [ 1, 3 ], " COMMON_VIDEO_CAPS "; "
@@ -1264,6 +1268,29 @@ skip_details:
       else if (!strcmp (fstr, "RGB"))
         videocontext->fourcc = GST_MAKE_FOURCC ('R', 'G', 'B', 24);
     }
+  } else if (!strcmp (mimetype, "video/rvl")) {
+    gint bpp=0;
+    const gchar *fstr;
+    gst_matroska_mux_set_codec_id (context, GST_MATROSKA_CODEC_ID_VIDEO_RVL);
+
+    gst_structure_get_int (structure, "bpp", &bpp);
+    switch(bpp) {
+      case 1:
+        videocontext->fourcc = GST_MAKE_FOURCC ('1', ' ', ' ', ' ');
+        break;
+      case 2:
+        videocontext->fourcc = GST_MAKE_FOURCC ('2', ' ', ' ', ' ');
+        break;
+      case 4:
+        videocontext->fourcc = GST_MAKE_FOURCC ('4', ' ', ' ', ' ');
+        break;
+      case 8:
+        videocontext->fourcc = GST_MAKE_FOURCC ('8', ' ', ' ', ' ');
+        break;
+      case 16:
+        videocontext->fourcc = GST_MAKE_FOURCC ('1', '6', ' ', ' ');
+        break;
+    }
   } else if (!strcmp (mimetype, "video/x-huffyuv")      /* MS/VfW compatibility cases */
       ||!strcmp (mimetype, "video/x-divx")
       || !strcmp (mimetype, "video/x-dv")
