diff --git a/subprojects/gst-plugins-good/sys/v4l2/v4l2_calls.c b/subprojects/gst-plugins-good/sys/v4l2/v4l2_calls.c
index 6243d09c39..c3c3af7649 100644
--- a/subprojects/gst-plugins-good/sys/v4l2/v4l2_calls.c
+++ b/subprojects/gst-plugins-good/sys/v4l2/v4l2_calls.c
@@ -365,6 +365,7 @@ gst_v4l2_fill_lists (GstV4l2Object * v4l2object)
       case V4L2_CID_AUTOGAIN:
       case V4L2_CID_GAIN:
       case V4L2_CID_SHARPNESS:
+      case V4L2_CID_MPEG_VIDEO_BITRATE:
         /* we only handle these for now (why?) */
         break;
       case V4L2_CID_HFLIP:
