diff --git a/net/aws/src/s3src/imp.rs b/net/aws/src/s3src/imp.rs
index 002bd612..651c0af7 100644
--- a/net/aws/src/s3src/imp.rs
+++ b/net/aws/src/s3src/imp.rs
@@ -159,10 +159,13 @@ impl S3Src {
                 settings.url = Some(s3url);
                 Ok(())
             }
-            Err(_) => Err(glib::Error::new(
-                gst::URIError::BadUri,
-                "Could not parse URI",
-            )),
+            Err(err) => {
+                gst::error!(CAT, "{}", err);
+                Err(glib::Error::new(
+                    gst::URIError::BadUri,
+                    "Could not parse URI",
+                ))
+            }
         }
     }
 
@@ -321,7 +324,9 @@ impl ObjectImpl for S3Src {
         match pspec.name() {
             "uri" => {
                 drop(settings);
-                let _ = self.set_uri(value.get().expect("type checked upstream"));
+                if let Err(err) = self.set_uri(value.get().expect("type checked upstream")) {
+                    gst::error!(CAT, "{}", err);
+                }
             }
             "access-key" => {
                 settings.access_key = value.get().expect("type checked upstream");
